zebra-network/src/protocol/external/addr/v1.rs

```diff
@@ -0,0 +1,169 @@
+//! Zcash `addr` (v1) message node address serialization.
+//!
+//! The [`AddrV1`] format serializes all IP addresses as IPv6 addresses.
+//! IPv4 addresses are converted to an [IPv4-mapped IPv6 address] before serialization.
+//!
+//! [IPv4-mapped IPv6 address]: https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
+
+use std::{
+    io::{Read, Write},
+    net::{IpAddr, Ipv6Addr, SocketAddr, SocketAddrV6},
+};
+
+use byteorder::{BigEndian, LittleEndian, ReadBytesExt, WriteBytesExt};
+
+use zebra_chain::serialization::{
+    DateTime32, SerializationError, TrustedPreallocate, ZcashDeserialize, ZcashDeserializeInto,
+    ZcashSerialize,
+};
+
+use crate::{
+    meta_addr::MetaAddr,
+    protocol::external::{types::PeerServices, MAX_PROTOCOL_MESSAGE_LEN},
+};
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest_derive::Arbitrary;
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use crate::protocol::external::arbitrary::addr_v1_ipv6_mapped_socket_addr_strategy;
+
+/// A version 1 node address, its advertised services, and last-seen time.
+/// This struct is serialized and deserialized into `addr` messages.
+///
+/// [Bitcoin reference](https://en.bitcoin.it/wiki/Protocol_documentation#Network_address)
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub(in super::super) struct AddrV1 {
+    /// The unverified "last seen time" gossiped by the remote peer that sent us
+    /// this address.
+    ///
+    /// See the [`MetaAddr::last_seen`] method for details.
+    untrusted_last_seen: DateTime32,
+
+    /// The unverified services for the peer at `ipv6_addr`.
+    ///
+    /// These services were advertised by the peer at `ipv6_addr`,
+    /// then gossiped via another peer.
+    ///
+    /// ## Security
+    ///
+    /// `untrusted_services` on gossiped peers may be invalid due to outdated
+    /// records, older peer versions, or buggy or malicious peers.
+    untrusted_services: PeerServices,
+
+    /// The peer's IPv6 socket address.
+    /// IPv4 addresses are serialized as an [IPv4-mapped IPv6 address].
+    ///
+    /// [IPv4-mapped IPv6 address]: https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
+    #[cfg_attr(
+        any(test, feature = "proptest-impl"),
+        proptest(strategy = "addr_v1_ipv6_mapped_socket_addr_strategy()")
+    )]
+    ipv6_addr: SocketAddrV6,
+}
+
+impl From<MetaAddr> for AddrV1 {
+    fn from(meta_addr: MetaAddr) -> Self {
+        let ipv6_addr = ipv6_mapped_socket_addr(meta_addr.addr);
+
+        let untrusted_services = meta_addr.services.expect(
```

Comment:
Perhaps an idea/discussion for later: should we have a `SanitizedMetaAddr` type? Or maybe have an `impl TryFrom<MetaAddr> for AddrV1` instead?

Reply:
A `SanitizedMetaAddr` type would be better, because we'd detect most bugs at compile time.
But it's out of scope for this refactor PR - we're focused on adding `addrv2`.

Can you open a ticket?

Reply:
Done: #3046.